eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/mutable/AbstractHashBag.java

```diff
@@ -373,4 +375,10 @@ public void remove()
             this.canRemove = false;
         }
     }
+
+    @Override
+    public MutableSet<T> asSet()
+    {
+        return SetAdapter.adapt(this.items.keySet());
```

Comment:
```java
@Test
public void asSet()
{
    MutableBag<Integer> bag = Bags.mutable.with(1, 2, 2, 3, 3, 3);
    MutableSet<Integer> set = bag.asSet();
    set.remove(3);
    Verify.assertSize(3, bag);
}
```

```
java.lang.AssertionError: Incorrect size for iterable; 
Expected :3
Actual   :6
```

Reply:
Ah, indeed, I didn't think about that at all! So yes, let's do that and stop trying to handle the mutable case. This can be done in another iteration I suppose

Reply:
One caveat with this is that if I do that, the return type of `MutableBag.asSet` is not a `MutableSet` anymore and I can't easily wrap all of the variants such as synchronized, unmodifiable and multireader without using some potentially unreliable casting...

Could we start by declaring `asSet` as returning a `MutableSet` but it would be an `UnmodifiableMutableSet` for now?

Reply:
SetIterable is more generic. You don't have to change anything about the implementations except that they also have to be unmodifiable.

Reply:
Sorry, I wasn't clear, my problem is in those kind of place: https://github.com/eclipse/eclipse-collections/pull/797/files#diff-bf399276403945a5c7592cb17e78051fR1081

As you can see, to be able to use `MultiReaderUnifiedSet.of(MutableSet)`, I need that `this.getDelegate().asSet()` returns a `MutableSet`, but in this case, the delegate is a `MutableBag`. So if `MutableBag.asSet()` returns only a `SetIterable`, I have a typing problem that I can only solve via casting, which is dangerous because I'm not sure that the implementation will actually return a `MutableSet`.

Do you see what I mean?

Reply:
In my last push, I've used the most precise interface possible everywhere except for the fact that `MutableSet` all returns unmodifiable ones in practice until the invariants can be fixed